### PR TITLE
Security groups can be referenced by group-name, don't need a data bag entry

### PR DESCRIPTION
**Pull request by tyler-ball**
fixes https://github.com/chef/chef-provisioning-aws/issues/49

\cc @jkeiser

### CODE REVIEW

File: lib/chef/resource/aws_security_group.rb [hunk 1]
```diff
@@ -49,8 +49,11 @@ class Chef::Resource::AwsSecurityGroup < Chef::Provisioning::AWSDriver::AWSResou
   }
 
   def aws_object
-    driver, id = get_driver_and_id
-    result = driver.ec2.security_groups[id] if id
+    if security_group_id
+      result = driver.ec2.security_groups[security_group_id]
+    else
+      result = driver.ec2.security_groups.filter('group-name', name).first
```
Comment (tyler-ball) [inline lib/chef/resource/aws_security_group.rb]: TODO if more than 1 comes back, we need to filter by VPC.  If the user does not provide VPC we blow up and say why.

**Comment by Joseph-R** [comment]
Confirmed working for me.  https://github.com/chef/chef-provisioning-aws/issues/174#issuecomment-98858430

File: lib/chef/resource/aws_security_group.rb [hunk 2]
```diff
@@ -1,8 +1,15 @@
 require 'chef/provisioning/aws_driver/aws_resource_with_entry'
 require 'chef/resource/aws_vpc'
 
-class Chef::Resource::AwsSecurityGroup < Chef::Provisioning::AWSDriver::AWSResourceWithEntry
-  aws_sdk_type AWS::EC2::SecurityGroup
+class MultipleSecurityGroupError < RuntimeError
```
Comment (tyler-ball) [inline lib/chef/resource/aws_security_group.rb]: TODO: extract this into `errors.rb` and have them descend from one base chef-provisioning-aws error.

File: lib/chef/resource/aws_security_group.rb [hunk 3]
```diff
@@ -49,8 +56,23 @@ class Chef::Resource::AwsSecurityGroup < Chef::Provisioning::AWSDriver::AWSResou
...
   def aws_object
-    driver, id = get_driver_and_id
-    result = driver.ec2.security_groups[id] if id
+    if security_group_id
+      result = driver.ec2.security_groups[security_group_id]
+    else
+      # Names are unique within a VPC.  Try to search by name and narroy by VPC, if
+      # provided
+      results = driver.ec2.security_groups.filter('group-name', name).to_a
+      if vpc
```
Comment (tyler-ball) [inline lib/chef/resource/aws_security_group.rb]: TODO: if you are provided the VPC, do `vpc.security_groups.filter`

File: spec/integration/aws_security_group_spec.rb [hunk 4]
```diff
@@ -1,56 +1,147 @@
...
+      it "correctly returns the security group when vpc is specified" do
+        aws_obj = nil
+        expect {
+          run_recipe {
+            r = aws_security_group 'test_sg' do
+              vpc 'test_vpc1'
+            end
+            aws_obj = r.aws_object
+          }
+        }.to_not raise_error
```
Comment (tyler-ball) [inline spec/integration/aws_security_group_spec.rb]: https://github.com/chef/chef/pull/3269/files#diff-ad16d58ada3f428b52eb88cb9e8b5b72R106

`emit_no_warnings_or_errors` or `expect_converge`

File: lib/chef/resource/aws_security_group.rb [hunk 5]
```diff
@@ -49,8 +56,23 @@ class Chef::Resource::AwsSecurityGroup < Chef::Provisioning::AWSDriver::AWSResou
...
+        vpc_object = Chef::Resource::AwsVpc.get_aws_object(vpc, resource: self)
+        results.select! {|sg| sg.vpc == vpc_object}
+      end
+      case results.size
```
Comment (tyler-ball) [inline lib/chef/resource/aws_security_group.rb]:
```ruby
if results.size >= 2
  raise MultipleSecurityGroupError.new(name, results)
end
result = results.first
```

**Pull request merged**